File: substrate/highlight_deep_research_launch_compose.py
"""Highlight → deep research launch package (pure).

live_dispatched and merge_executed always False.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Literal

from substrate.floating_deep_research import (
    FloatingDeepResearchError,
    FloatingDeepResearchInstance,
    spawn_floating_from_highlight,
)

LaunchPreferredView = Literal["floating", "fullscreen"]
SourceFamilyHint = Literal["arxiv", "substack", "openalex", "web", "custom"]
VALID_FAMILIES = frozenset(
    ("arxiv", "substack", "openalex", "web", "custom")
)
_SECRETISH = re.compile(r"sk-|api[_-]?key|secret", re.I)


class HighlightDeepResearchLaunchComposeError(ValueError):
    """Fail-closed validation for highlight DR launch."""


@dataclass(frozen=True)
class HighlightDeepResearchLaunchCompose:
    instance: FloatingDeepResearchInstance
    preferred_view_mode: str
    selected_model_id: str | None
    source_families: tuple[str, ...]
    source_family_count: int
    budget_ready: bool
    would_exceed: bool | None
    launch_ready: bool
    live_dispatched: bool
    merge_executed: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "instance": self.instance.to_dict(),
            "preferred_view_mode": self.preferred_view_mode,
            "selected_model_id": self.selected_model_id,
            "source_families": list(self.source_families),
            "source_family_count": self.source_family_count,
            "budget_ready": self.budget_ready,
            "would_exceed": self.would_exceed,
            "launch_ready": self.launch_ready,
            "live_dispatched": False,
            "merge_executed": False,
            "notes": list(self.notes),
            "authority": "highlight_deep_research_launch_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise HighlightDeepResearchLaunchComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_highlight_deep_research_launch(
    *,
    parent_asset_id: object,
    highlight: object,
    gated: object,
    would_exceed: object,
    operator_ack: object,
    prompt: object | None = None,
    preferred_view_mode: object | None = None,
    operator_override: object | None = None,
    selected_model_id: object | None = None,
    source_families: object | None = None,
) -> HighlightDeepResearchLaunchCompose:
    """Compose pure DR launch from highlight. Never live-dispatches."""
    if not isinstance(operator_ack, bool):
        raise HighlightDeepResearchLaunchComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(gated, bool):
        raise HighlightDeepResearchLaunchComposeError(
            "gated must be an explicit boolean from highlight provenance (fail closed)"
        )
    if would_exceed is not None and not isinstance(would_exceed, bool):
        raise HighlightDeepResearchLaunchComposeError(
            "would_exceed must be boolean or null"
        )
    override = False if operator_override is None else operator_override
    if not isinstance(override, bool):
        raise HighlightDeepResearchLaunchComposeError(
            "operator_override must be boolean when set"
        )

    view: LaunchPreferredView = "floating"
    if preferred_view_mode is not None:
        if preferred_view_mode not in ("floating", "fullscreen"):
            raise HighlightDeepResearchLaunchComposeError(
                "preferred_view_mode must be floating|fullscreen"
            )
        view = preferred_view_mode  # type: ignore[assignment]

    notes: list[str] = [
        "live_dispatched=false — launch package is pure intent only",
        "merge_executed=false — parent asset not mutated",
    ]

    try:
        instance = spawn_floating_from_highlight(
            parent_asset_id=parent_asset_id,
            highlight=highlight,
            gated=gated,
            prompt=prompt,
            view_mode=view,
        )
    except FloatingDeepResearchError as e:
        raise HighlightDeepResearchLaunchComposeError(str(e)) from e

    notes.append(
        f"spawned instance_id={instance.instance_id} · view_mode={instance.view_mode}"
    )
    notes.append("live_dispatched=false on instance")

    model_id: str | None = None
    if selected_model_id is not None:
        model_id = _require_nonempty(
            selected_model_id, field="selected_model_id"
        )
        if len(model_id) > 128 or _SECRETISH.search(model_id):
            raise HighlightDeepResearchLaunchComposeError(
                "selected_model_id must be a model id, not secret material"
            )
        notes.append(f"selected_model_id={model_id} (operator authority)")
    else:
        notes.append(
            "selected_model_id=null — operator may choose before live launch"
        )

    families: list[str] = []
    if source_families is not None:
        if not isinstance(source_families, list):
            raise HighlightDeepResearchLaunchComposeError(
                "source_families must be an array when set"
            )
        seen: set[str] = set()
        for i, f in enumerate(source_families):
            if f not in VALID_FAMILIES:
                raise HighlightDeepResearchLaunchComposeError(
                    f"source_families[{i}] must be arxiv|substack|openalex|web|custom"
                )
            if f in seen:
                raise HighlightDeepResearchLaunchComposeError(
                    f"duplicate source_family: {f}"
                )
            seen.add(str(f))
            families.append(str(f))
    notes.append(f"source_family_count={len(families)}")

    budget_ready = False
    if would_exceed is None:
        if override:
            budget_ready = True
            notes.append(
                "budget_ready=true via operator_override (would_exceed unknown)"
            )
        else:
            notes.append(
                "budget_ready=false — would_exceed unknown and no operator_override"
            )
    elif would_exceed is True:
        if override:
            budget_ready = True
            notes.append(
                "budget_ready=true via operator_override despite would_exceed=true"
            )
        else:
            notes.append("budget_ready=false — would_exceed=true")
    else:
        budget_ready = True
        notes.append("budget_ready=true — would_exceed=false")

    launch_ready = (
        operator_ack
        and budget_ready
        and instance.live_dispatched is False
        and instance.merge_executed is False
    )
    if not operator_ack:
        notes.append("launch_ready=false — operator_ack required")
    elif not budget_ready:
        notes.append("launch_ready=false — budget gate closed")
    else:
        notes.append(
            "launch_ready=true — pure package ready; still live_dispatched=false"
        )
    notes.extend(("live_dispatched=false", "merge_executed=false"))

    return HighlightDeepResearchLaunchCompose(
        instance=instance,
        preferred_view_mode=view,
        selected_model_id=model_id,
        source_families=tuple(families),
        source_family_count=len(families),
        budget_ready=budget_ready,
        would_exceed=would_exceed if isinstance(would_exceed, bool) or would_exceed is None else None,
        launch_ready=launch_ready,
        live_dispatched=False,
        merge_executed=False,
        notes=tuple(notes),
        authority="highlight_deep_research_launch_compose_advisory",
    )


__all__ = [
    "HighlightDeepResearchLaunchCompose",
    "HighlightDeepResearchLaunchComposeError",
    "compose_highlight_deep_research_launch",
]

File: apps/reading/src/api/highlightDeepResearchLaunchCompose.ts
/**
 * Highlight → deep research launch package (pure).
 *
 * Operator vision: from a reading/research highlight, spin up floating deep
 * research; choose float/fullscreen; know budget impact; optionally attach
 * arxiv/substack source families — without live dispatch.
 *
 * live_dispatched always false.
 * merge_executed always false.
 */

import {
  spawnFloatingFromHighlight,
  type FloatingDeepResearchInstance,
  type FloatingResearchViewMode,
} from "./floatingDeepResearch";

export type LaunchPreferredView = "floating" | "fullscreen";

export type SourceFamilyHint =
  | "arxiv"
  | "substack"
  | "openalex"
  | "web"
  | "custom";

export interface HighlightDeepResearchLaunchInput {
  parent_asset_id: string;
  highlight: string;
  /** Explicit gated flag from highlight provenance — fail closed if true. */
  gated: boolean;
  /** Optional operator prompt override. */
  prompt?: string;
  preferred_view_mode?: LaunchPreferredView;
  /**
   * Budget would_exceed for the proposed DR prompt; null = unknown honesty.
   */
  would_exceed: boolean | null;
  /** Operator override when budget unknown or would_exceed. */
  operator_override?: boolean;
  /** Optional selected model id (operator authority). */
  selected_model_id?: string | null;
  /** Optional source families to attach for this launch. */
  source_families?: SourceFamilyHint[] | null;
  operator_ack: boolean;
}

export interface HighlightDeepResearchLaunchCompose {
  instance: FloatingDeepResearchInstance;
  preferred_view_mode: LaunchPreferredView;
  selected_model_id: string | null;
  source_families: SourceFamilyHint[];
  source_family_count: number;
  budget_ready: boolean;
  would_exceed: boolean | null;
  /**
   * True when spawn succeeded, operator_ack, budget_ready, not gated.
   * Still never live-dispatches.
   */
  launch_ready: boolean;
  /** Always false — pure layer never dispatches providers. */
  live_dispatched: false;
  /** Always false — pure layer never merges into parent. */
  merge_executed: false;
  notes: string[];
  authority: "highlight_deep_research_launch_compose_advisory";
}

const VALID_FAMILIES = new Set<SourceFamilyHint>([
  "arxiv",
  "substack",
  "openalex",
  "web",
  "custom",
]);

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose a pure deep-research launch package from a highlight.
 * Never live-dispatches; never merges.
 */
export function composeHighlightDeepResearchLaunch(
  input: HighlightDeepResearchLaunchInput,
): HighlightDeepResearchLaunchCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (typeof input.gated !== "boolean") {
    throw new Error(
      "gated must be an explicit boolean from highlight provenance (fail closed)",
    );
  }
  if (
    input.would_exceed !== null &&
    input.would_exceed !== undefined &&
    typeof input.would_exceed !== "boolean"
  ) {
    throw new Error("would_exceed must be boolean or null");
  }
  const would_exceed =
    input.would_exceed === undefined ? null : input.would_exceed;

  const override =
    input.operator_override === undefined ? false : input.operator_override;
  if (typeof override !== "boolean") {
    throw new Error("operator_override must be boolean when set");
  }

  let preferred_view_mode: LaunchPreferredView = "floating";
  if (input.preferred_view_mode != null && input.preferred_view_mode !== undefined) {
    if (
      input.preferred_view_mode !== "floating" &&
      input.preferred_view_mode !== "fullscreen"
    ) {
      throw new Error("preferred_view_mode must be floating|fullscreen");
    }
    preferred_view_mode = input.preferred_view_mode;
  }

  const notes: string[] = [
    "live_dispatched=false — launch package is pure intent only",
    "merge_executed=false — parent asset not mutated",
  ];

  // Spawn pure floating instance (reuses existing fail-closed helpers).
  const instance = spawnFloatingFromHighlight({
    parent_asset_id: input.parent_asset_id,
    highlight: input.highlight,
    gated: input.gated,
    prompt: input.prompt,
    view_mode: preferred_view_mode as FloatingResearchViewMode,
  });
  notes.push(
    `spawned instance_id=${instance.instance_id} · view_mode=${instance.view_mode}`,
  );
  notes.push("live_dispatched=false on instance");

  let selected_model_id: string | null = null;
  if (
    input.selected_model_id != null &&
    input.selected_model_id !== undefined
  ) {
    selected_model_id = requireNonEmpty(
      input.selected_model_id,
      "selected_model_id",
    );
    if (
      selected_model_id.length > 128 ||
      /sk-|api[_-]?key|secret/i.test(selected_model_id)
    ) {
      throw new Error(
        "selected_model_id must be a model id, not secret material",
      );
    }
    notes.push(`selected_model_id=${selected_model_id} (operator authority)`);
  } else {
    notes.push("selected_model_id=null — operator may choose before live launch");
  }

  const source_families: SourceFamilyHint[] = [];
  if (input.source_families != null) {
    if (!Array.isArray(input.source_families)) {
      throw new Error("source_families must be an array when set");
    }
    const seen = new Set<string>();
    for (let i = 0; i < input.source_families.length; i++) {
      const f = input.source_families[i];
      if (!VALID_FAMILIES.has(f)) {
        throw new Error(
          `source_families[${i}] must be arxiv|substack|openalex|web|custom`,
        );
      }
      if (seen.has(f)) {
        throw new Error(`duplicate source_family: ${f}`);
      }
      seen.add(f);
      source_families.push(f);
    }
  }
  notes.push(`source_family_count=${source_families.length}`);

  let budget_ready = false;
  if (would_exceed === null) {
    if (override) {
      budget_ready = true;
      notes.push(
        "budget_ready=true via operator_override (would_exceed unknown)",
      );
    } else {
      notes.push(
        "budget_ready=false — would_exceed unknown and no operator_override",
      );
    }
  } else if (would_exceed === true) {
    if (override) {
      budget_ready = true;
      notes.push(
        "budget_ready=true via operator_override despite would_exceed=true",
      );
    } else {
      notes.push("budget_ready=false — would_exceed=true");
    }
  } else {
    budget_ready = true;
    notes.push("budget_ready=true — would_exceed=false");
  }

  const launch_ready =
    input.operator_ack &&
    budget_ready &&
    instance.live_dispatched === false &&
    instance.merge_executed === false;

  if (!input.operator_ack) {
    notes.push("launch_ready=false — operator_ack required");
  } else if (!budget_ready) {
    notes.push("launch_ready=false — budget gate closed");
  } else {
    notes.push(
      "launch_ready=true — pure package ready; still live_dispatched=false",
    );
  }

  notes.push("live_dispatched=false");
  notes.push("merge_executed=false");

  return {
    instance,
    preferred_view_mode,
    selected_model_id,
    source_families,
    source_family_count: source_families.length,
    budget_ready,
    would_exceed,
    launch_ready,
    live_dispatched: false,
    merge_executed: false,
    notes,
    authority: "highlight_deep_research_launch_compose_advisory",
  };
}

export function formatHighlightDeepResearchLaunchSummary(
  c: HighlightDeepResearchLaunchCompose,
): string {
  return (
    `launch_ready=${c.launch_ready} · mode=${c.preferred_view_mode} · ` +
    `sources=${c.source_family_count} · budget_ready=${c.budget_ready} · ` +
    `live_dispatched=false · merge_executed=false`
  );
}
